pkg/netpol/eval/resources.go

```diff
@@ -574,48 +585,76 @@ func (pe *PolicyEngine) GetSelectedPeers(selectors labels.Selector, namespace st
-		if _, ok := pe.representativePeersMap[keyStrFromLabels]; ok { // we already have a representative peer with same labels
-			return nil, nil
-		}
-		newRepresentativePeer := &k8s.RepresentativePeer{Pod: newPod, PotentialNamespaceLabels: nsLabels}
-		pe.representativePeersMap[keyStrFromLabels] = newRepresentativePeer
-		return newRepresentativePeer, nil
-	}
-	// ingress-controller will be treated as a real pod, may be added to podsMap
 	pe.podsMap[podStr] = newPod
 	return &k8s.WorkloadPeer{Pod: newPod}, nil
 }
+
+// addRepresentativePod adds a new representative pod to the policy-engine (to pe.representativePeersMap).
+// if the given namespace string (podNs) is not empty (i.e. a real (policy's) namespace name), it will be assigned to the pod's Namespace;
+// and the "namespace name" requirement of the representative pod will be stored in its RepresentativeNsLabelSelector field.
+// Otherwise, the representative pod will have no namespace (will not add a representative namespace to the policy-engine).
+// this func is used only with exposure-analysis
+func (pe *PolicyEngine) addRepresentativePod(podNs string, objSelectors *k8s.SingleRuleSelectors) error {
+	if objSelectors == nil { // should not get here
+		return errors.New(netpolerrors.NilRepresentativePodSelectorsErr)
+	}
+	nsLabelSelector := objSelectors.NsSelector
+	if nsLabelSelector == nil { // equal to podNs != ""
+		// if the objSelectors.NsSelector is nil, means inferred from a rule with nil nsSelector, which means the namespace of the
+		// pod is the namespace of the policy, so adding it as its RepresentativeNsLabelSelector requirement.
+		// by this, we ensure a representative peer may only represent the rule it was inferred from
+		// and uniqueness of representative peers.
+		// (another policy in another namespace, may have a rule with same podSelector, but the namespace will be different-
+		// so a different representative peer will be generated)
+		nsLabelSelector = &metav1.LabelSelector{MatchLabels: defaultNamespaceLabelsMap(podNs)}
+	}
+	newPod := &k8s.Pod{
+		// all representative pods are having same name since this name is used only to indicate that this Fake Pod is representative;
+		// this name is not used for storing it in the policy-engine/ comparing with other peers/ or representing it.
+		Name:                           k8s.RepresentativePodName,
+		Namespace:                      podNs,
+		FakePod:                        true,
+		RepresentativePodLabelSelector: objSelectors.PodSelector,
+		RepresentativeNsLabelSelector:  nsLabelSelector,
+	}
+	//  compute a unique string from the label selectors to be used as the map key
+	// note that nsLabelSelector will not be nil
+	nsKey, err := k8s.VariantFromLabelsSelector(nsLabelSelector)
+	if err != nil {
+		return err
+	}
+	podKey, err := k8s.VariantFromLabelsSelector(objSelectors.PodSelector)
+	if err != nil {
+		return err
+	}
+	keyStrFromLabels := nsKey + "/" + podKey
```

Comment:
what if one policy has nil ns selector and pod selecotr {app: foo}, within namespace N1 [N1 is defined with label {x: xval}],
and another policy has ns selector {x: xval} and pod selecotr {app: foo} ?

Reply:
- if there is a real pod in `N1` then the representative peers inferred from both policy rules will be removed (refined) and not appear in the output.
- if there is no real pod in the ns, then we will see 2 representative peers; a representative peer in `N1` and another representative-peer in any-ns with {x:val} >> since we refine rep-peers matching pods; not matching namespaces.
- added tests